Prometheus/TextSerializer.cs

```diff
@@ -49,19 +56,77 @@ public async Task WriteFamilyDeclarationAsync(byte[][] headerLines, Cancellation
         // Size limit guided by https://stackoverflow.com/questions/21146544/what-is-the-maximum-length-of-double-tostringd
         private readonly byte[] _stringBytesBuffer = new byte[32];
 
-        // name{labelkey1="labelvalue1",labelkey2="labelvalue2"} 123.456
-        public async Task WriteMetricAsync(byte[] identifier, double value, CancellationToken cancel)
+        // 123.456
+        // Note: Terminates with a NEWLINE
+        public async Task WriteValuePartAsync(double value, CancellationToken cancel)
         {
-            await _stream.Value.WriteAsync(identifier, 0, identifier.Length, cancel);
-            await _stream.Value.WriteAsync(Space, 0, Space.Length, cancel);
-
             var valueAsString = value.ToString(CultureInfo.InvariantCulture);
-
             var numBytes = PrometheusConstants.ExportEncoding
                 .GetBytes(valueAsString, 0, valueAsString.Length, _stringBytesBuffer, 0);
 
             await _stream.Value.WriteAsync(_stringBytesBuffer, 0, numBytes, cancel);
             await _stream.Value.WriteAsync(NewLine, 0, NewLine.Length, cancel);
         }
+
+        /// <summary>
+        /// Creates a metric identifier, with an optional name postfix and an optional extra label to append to the end.
+        /// familyname_postfix{labelkey1="labelvalue1",labelkey2="labelvalue2"}
+        /// Note: Terminates with a SPACE
+        /// </summary>
+        public async Task WriteIdentifierPartAsync(byte[] name, byte[] flatennedLabels, CancellationToken cancel,
+            byte[]? postfix = null, byte[]? extraLabelName = null, byte[]? extraLabelValue = null,
+            byte[]? extraLabelValueOpenMetrics = null)
+        {
+            await _stream.Value.WriteAsync(name, 0, name.Length, cancel);
```

Comment:
There was an issue reported to prometheus-net repo a while ago that under ??? conditions the reporter saw that there was performance degradation from the many WriteAsync() calls. They never specified how to reproduce, though, so I do not know the exact situation.

There are now many more WriteAsync calls. I understand you ran the benchmarks but I wonder if there will be some problem that is not exposed by the benchmarks (e.g. because of MemoryStream versus network stream differences). Thoughts?

Reply:
@sandersaares I too am concerned about this (ref: out of band discussion). 

I propose we tackle this towards the end of the feature. 

If the PR looks good perhaps we can have a staging branch ? I envisage two more PRs to feature completion.

1. OpenMetrics encoding support -- minimal, aiming for feature parity with golang -- with tests.
2. HTTP Protocol negotiation function and make OpenMetrics configurable as a protocol programatically.

Reply:
Looks good. Agree to worry about async at a later point. I will merge this PR and we can iterate, either directly into master branch or to a separate branch if the next iteration is not merge-ready yet.